hkmc2/shared/src/test/mlscript/handlers/StackSafety.mls

```diff
@@ -1,76 +1,59 @@
 :js
+:effectHandlers
+:lift
+:noSanityCheck
 :noTailRec
 
 // * FIXME: Why doesn't the following work when using Predef function `(==) equals`?
 
-// sanity check
-:expect 5050
-fun sum(n) =
-  if n == 0 then 0
-  else
-    n + sum(n - 1)
-sum(100)
-//│ = 5050
 
 // preserve tail calls
 // MUST see "return hi(tmp)" in the output
 :stackSafe 5
-:effectHandlers
 :expect 0
 :sjs
 fun hi(n) =
   if n === 0 then 0
   else hi(n - 1)
 hi(0)
 //│ JS (unsanitized):
-//│ let hi;
-//│ let res, $_stack$_safe$_body$_;
+//│ let hi, tmp;
+//│ let $_stack$_safe$_body$_;
 //│ hi = function hi(n) {
-//│   let scrut, tmp;
-//│   let Cont$func$hi$1, doUnwind, stackDelayRes;
-//│   (class Cont$func$hi$ extends runtime.FunctionContFrame.class {
-//│     static {
-//│       Cont$func$hi$1 = this
-//│     }
-//│     constructor(pc) {
-//│       let tmp1;
-//│       tmp1 = super(null);
-//│       this.pc = pc;
-//│     }
-//│     resume(value$) {
-//│       return hi(n)
-//│     }
-//│     toString() { return runtime.render(this); }
-//│     static [definitionMetadata] = ["class", "Cont$func$hi$"]; 
-//│   });
-//│   doUnwind = function doUnwind(res1, pc) {
-//│     res1.contTrace.last.next = new Cont$func$hi$1(pc);
-//│     res1.contTrace.last = res1.contTrace.last.next;
-//│     return res1
-//│   };
+//│   let scrut, tmp1, pc;
+//│   let stackDelayRes;
 //│   runtime.stackDepth = runtime.stackDepth + 1;
 //│   stackDelayRes = runtime.checkDepth();
 //│   if (stackDelayRes instanceof runtime.EffectSig.class) {
-//│     return doUnwind(stackDelayRes, 0)
+//│     return runtime.unwind(stackDelayRes, 1, hi, null, "StackSafety.mls:15:1", 0, null, 2, tmp1, n)
+//│   }
+//│   if (runtime.resumePc === -1) {
+//│     pc = 0;
+//│   } else {
+//│     let saveOffset;
+//│     pc = runtime.resumePc;
+//│     tmp1 = runtime.resumeArr.at(runtime.resumeIdx);
```

Comment:
Do you think it would be feasible to avoid restoring the values of variables like `tmp1` here, since it is always undefined at the points where the function suspends?

In fact,
* Aren't all variables always undefined for `unwind` calls caused by stack safety?
* More generally, could we do a quick/simple variable liveness analysis of the `unwind` call to only include variables that are live for at least one of these calls?

This could be done in a follow-up PR, though.

Reply:
parameters are also treated as local variables currently. Actually I've switch my mind a bit and I think callling the function with the correct parameter list might be worth it since we can avoid special handling of stack safety in the generated code when there is one state.